### PR TITLE
5498 file hierarchy in zip downloads

### PR DESCRIPTION
**Pull request by landreev**
## New Contributors

Welcome! New contributors should at least glance at [CONTRIBUTING.md](/CONTRIBUTING.md), especially the section on pull requests where we encourage you to reach out to other developers before you start coding. Also, please note that we measure code coverage and prefer you write unit tests. Pull requests can still be reviewed without tests or completion of the checklist outlined below. Thanks!

## Related Issues

- connects to #5498 : file hierarchy in zip downloads

## Pull Request Checklist

- [ ] Unit [tests][] completed
- [ ] Integration [tests][]: None
- [ ] Deployment requirements, [SQL updates][], [Solr updates][], etc.: None
- [ ] [Documentation][docs] completed
- [ ] Merged latest from "develop" [branch][] and resolved conflicts

[tests]: http://guides.dataverse.org/en/latest/developers/testing.html
[SQL updates]: http://guides.dataverse.org/en/latest/developers/sql-upgrade-scripts.html
[Solr updates]: https://github.com/IQSS/dataverse/blob/develop/conf/solr/7.3.0/schema.xml
[docs]: http://guides.dataverse.org/en/latest/developers/documentation.html
[branch]: http://guides.dataverse.org/en/latest/developers/branching-strategy.html

### CODE REVIEW

File: src/main/java/edu/harvard/iq/dataverse/dataaccess/DataFileZipper.java [hunk 1]
```diff
@@ -133,12 +138,26 @@ public long addFileToZipStream(DataFile dataFile, boolean getOriginal) throws IO
 
                 Success = false;
             } else {
+                String folderName = dataFile.getFileMetadata().getDirectoryLabel(); 
+                if (folderName != null) {
+                    while (folderName.startsWith("/")) {
```
Comment (matthew-a-dunlap) [inline src/main/java/edu/harvard/iq/dataverse/dataaccess/DataFileZipper.java]: I'm a bit confused by this section. Does this preserve the nested folders? It seems like this strips all the folders until the last one.

A comment on what this section is doing could be useful for future maintenance.

File: src/main/java/edu/harvard/iq/dataverse/util/FileUtil.java [hunk 2]
```diff
@@ -543,10 +543,28 @@ public static String CalculateChecksum(InputStream in, ChecksumType checksumType
             }
         }
 
-        byte[] mdbytes = md.digest();
+        return checksumDigestToString(md.digest());
+    }
+    
+    public static String CalculateChecksum(byte[] dataBytes, ChecksumType checksumType) {
```
Comment (matthew-a-dunlap) [inline src/main/java/edu/harvard/iq/dataverse/util/FileUtil.java]: Camel-case?

File: src/main/java/edu/harvard/iq/dataverse/dataaccess/DataFileZipper.java (hunk 1 above)
Comment (landreev) [reply src/main/java/edu/harvard/iq/dataverse/dataaccess/DataFileZipper.java]: No, it only strips the leading slashes, if there are any. "///foo/bar" will become "foo/bar". 
I'll add a comment yes.

File: src/main/java/edu/harvard/iq/dataverse/util/FileUtil.java (hunk 2 above)
Comment (matthew-a-dunlap) [reply src/main/java/edu/harvard/iq/dataverse/util/FileUtil.java]: I think the method should be `calculateChecksum()`

File: src/main/java/edu/harvard/iq/dataverse/dataaccess/DataFileZipper.java (hunk 1 above)
Comment (matthew-a-dunlap) [reply src/main/java/edu/harvard/iq/dataverse/dataaccess/DataFileZipper.java]: Ah ok that makes sense! I was getting substring() mixed up with split()

File: src/main/java/edu/harvard/iq/dataverse/util/FileUtil.java (hunk 2 above)
Comment (landreev) [reply src/main/java/edu/harvard/iq/dataverse/util/FileUtil.java]: got it. I just used the name of the existing methods. 
But I have renamed all three - done.

File: src/main/java/edu/harvard/iq/dataverse/dataset/DatasetUtil.java [hunk 3]
```diff
@@ -34,6 +34,13 @@
 import java.util.Map;
 import javax.imageio.ImageIO;
 import org.apache.commons.io.IOUtils;
+import static edu.harvard.iq.dataverse.dataaccess.DataAccess.getStorageIO;
```
Comment (matthew-a-dunlap) [inline src/main/java/edu/harvard/iq/dataverse/dataset/DatasetUtil.java]: Looks like glassfish went a bit nuts doing a refactor

File: src/main/java/edu/harvard/iq/dataverse/util/FileUtil.java [hunk 4]
```diff
@@ -77,6 +77,13 @@
 import java.util.zip.GZIPInputStream;
 import java.util.zip.ZipEntry;
 import java.util.zip.ZipInputStream;
+import static edu.harvard.iq.dataverse.datasetutility.FileSizeChecker.bytesToHumanReadable;
```
Comment (matthew-a-dunlap) [inline src/main/java/edu/harvard/iq/dataverse/util/FileUtil.java]: Looks like glassfish went a bit nuts doing a refactor

File: src/test/java/edu/harvard/iq/dataverse/api/AccessIT.java [hunk 5]
```diff
@@ -31,6 +31,12 @@
 import static org.junit.Assert.assertNotNull;
 import static org.junit.Assert.assertThat;
 import static org.junit.Assert.assertTrue;
+import static org.hamcrest.CoreMatchers.is;
```
Comment (matthew-a-dunlap) [inline src/test/java/edu/harvard/iq/dataverse/api/AccessIT.java]: Looks like glassfish went a bit nuts doing a refactor

File: src/main/java/edu/harvard/iq/dataverse/api/Access.java [hunk 6]
```diff
@@ -107,6 +107,13 @@
 import javax.ws.rs.core.Response;
 import static javax.ws.rs.core.Response.Status.BAD_REQUEST;
 import javax.ws.rs.core.StreamingOutput;
+import static edu.harvard.iq.dataverse.util.json.JsonPrinter.json;
```
Comment (matthew-a-dunlap) [inline src/main/java/edu/harvard/iq/dataverse/api/Access.java]: Looks like glassfish went a bit nuts doing a refactor

Comment (landreev) [reply src/main/java/edu/harvard/iq/dataverse/api/Access.java]: you mean git... admittedly, I didn't look - just checked in all the modified files. All it was supposed to do was change all the references to CalculateChecksum to calculcateChecksum. 
(don't fix what ain't broke, lol)

Comment (landreev) [reply src/main/java/edu/harvard/iq/dataverse/api/Access.java]: huh - I removed a whole bunch of these unused imports; it looks like they were accumulating there (added by past git refactorings?)

**Pull request merged**